### DISCUSSION TITLE
How to setup python formatting?

### DISCUSSION THREAD

**Discussion by MrPandey01** [VonHeikemen/lsp-zero.nvim#50] [Q&A] [answered]
It turns out that Pyright language server does not support formatting. I later installed black for formatting (using mason). 

How do I configure lsp-zero such that the buffer is formatted when `vim.lsp.buf.format()` is called?

Currently this command throws an error that 'no LSP server is attached'.

**Comment by VonHeikemen** [comment, marked as answer]
You can use [null-ls](https://github.com/jose-elias-alvarez/null-ls.nvim). They have support for black: [builtins - black](https://github.com/jose-elias-alvarez/null-ls.nvim/blob/main/doc/BUILTINS.md#black).

Here is an example configuration.

```lua
local lsp = require('lsp-zero')
lsp.preset('recommended')

lsp.setup()

local null_ls = require('null-ls')
local null_opts = lsp.build_options('null-ls', {})

null_ls.setup({
  on_attach = function(client, bufnr)
    null_opts.on_attach(client, bufnr)
    --- you can add more stuff here if you need it
  end,
  sources = {
    null_ls.builtins.formatting.black,
  }
})
```

**Reply by MrPandey01** [reply]
Thanks! Works like charm :)

**Comment by kskarthik** [comment]
@VonHeikemen just got here to say thank you for lsp-zero plugin! It saved a lot of initial lsp setup for me!

**Reply by VonHeikemen** [reply]
I'm glad it was helpful.

**Comment by kevinsoewondo** [comment]
Amazing plugin, thank you so much! With so many different tutorials on lsp config, this plugin has saved so much time.

**Reply by VonHeikemen** [reply]
Awesome. Thank you for the kind words.

**Reply by Am3ra** [reply]
Thanks a lot @VonHeikemen ! Made my setup with lsps super simple. Now i wanna figure out how to integrate copilot, but it's already working well separately, so maybe that's for the weekend!

Great stuff, thanks for the hard work
